tls/s2n_ktls.c

```diff
@@ -15,11 +15,176 @@
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE(cipher->ktls_supported, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE_REF(conn->send_io_context);
+            RESULT_ENSURE(conn->managed_send_io, S2N_ERR_KTLS_MANAGED_IO);
+
+            /* The output stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE_REF(conn->recv_io_context);
+            RESULT_ENSURE(conn->managed_recv_io, S2N_ERR_KTLS_MANAGED_IO);
+
+            /* The input stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        RESULT_ENSURE_REF(conn->recv_io_context);
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        /* retrieve the send fd */
+        RESULT_ENSURE_REF(conn->send_io_context);
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* If already enabled then return success */
+    if (ktls_mode == S2N_KTLS_MODE_SEND && conn->ktls_send_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+    if (ktls_mode == S2N_KTLS_MODE_RECV && conn->ktls_recv_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
+    RESULT_ENSURE(!disable_ktls_socket_config_for_testing, S2N_ERR_KTLS_DISABLED_FOR_TEST);
+
+    /* Enable 'tls' ULP for the socket. https://lwn.net/Articles/730207 */
+    int ret = setsockopt(fd, S2N_SOL_TCP, S2N_TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE);
+
+    if (ret != 0) {
+        /* EEXIST: https://man7.org/linux/man-pages/man3/errno.3.html
+         *
+         * TCP_ULP has already been enabled on the socket so the operation is a noop.
```

Comment:
But I find this comment really confusing. I think I would phrase it as
"The EEXIST errno signifies the option TCP_ULP has already been enabled on the socket. This is the only setsockopt error that can be safely ignored."
But I could be wrong about what you're trying to say so feel free to correct me!

Reply:
updated in https://github.com/aws/s2n-tls/pull/4066/commits/c21995a2e53bbced297b6560743a50184445d495